src/Movement/Kinematics/HangprinterKinematics.cpp

```diff
@@ -330,17 +433,26 @@ static bool isSameSide(float const v0[3], float const v1[3], float const v2[3],
 	return dot0*dot1 > 0.0F;
 }
 
-static bool isInsideTetrahedron(float const point[3], float const tetrahedron[4][3]){
-	return isSameSide(tetrahedron[0], tetrahedron[1], tetrahedron[2], tetrahedron[3], point) &&
-	       isSameSide(tetrahedron[2], tetrahedron[1], tetrahedron[3], tetrahedron[0], point) &&
-	       isSameSide(tetrahedron[2], tetrahedron[3], tetrahedron[0], tetrahedron[1], point) &&
-	       isSameSide(tetrahedron[0], tetrahedron[3], tetrahedron[1], tetrahedron[2], point);
-}
-
 bool HangprinterKinematics::IsReachable(float axesCoords[MaxAxes], AxesBitmap axes) const noexcept /*override*/
 {
 	float const coords[3] = {axesCoords[X_AXIS], axesCoords[Y_AXIS], axesCoords[Z_AXIS]};
-	return isInsideTetrahedron(coords, anchors);
+
+	bool reachable = isSameSide(anchors[0], anchors[1], anchors[2], anchors[numAnchors - 1], coords);
+
+	for (size_t i = 1; reachable && i < numAnchors - 3; ++i)
+	{
+		reachable = isSameSide(anchors[i], anchors[i + 1], anchors[numAnchors - 1], anchors[i + 2], coords);
```

Comment:
This line of code divides anchors into
First anchors: (i) and (i+1)
Last anchor: (numAnchors - 1)
Mid anchor: (i + 2)
It stipulates that if mid anchor is on one side of the triangle between first+last anchors, then coord must be on the same side.
This requires the user to put first anchors in either clockwise or anticlockwise order around an axis going through last anchor, I think. If I'm right then anchor placement becomes not really arbitrary anymore.
Anyhow, this function needs a big battery of tests since it's hard to follow and reson about.

If we really want a function that works for any convex reachable volume, then we need to take the set of vertices (anchors), construct all triangles that span the convex outer surface (similar to what loft() in cad programs does), and for each triangle check that coord is on the same side as the non-triangle vertices are.

Reply:
To allow effectors to travel below the spanned volume, as I know you want to, it would be useful to run a 2d-version of the algorithm first, and if coord is below all anchors (but not too far into negative z), then the point is reachable.

Reply:
To sort a set of 2d-points, and extract the convex hull of them, you can do like line-collision-detector does, here: https://gitlab.com/hangprinter/linc/-/blob/master/linc/linc.c%2B%2B#L47

Reply:
TOREAD

Reply:
What I assumed isSameSide does:
You provide 3 points defining a plane.
You provide 2 additional points.
The plane divides space in two. if the two additional points are on the same side of that bisection of space, it returns true and otherwise it returns false.
I'm trying to be precise, but perhaps I'm losing on clarity.
But I just assumed that without even trying to understand the math inside the function, just by looking at how it was being used and because that made sense to me.

Based on that, what I'm trying to do here still assumes a pyramid configuration, like in the hangprinter.  
Just allows pyramids with one extra side (one extra triangle) for each extra anchor.
This assumes the last anchor is the one on the top of the pyramid (which I think it's already an assumption with the current 4-anchors implementation).
So what this supposedly does is going through all the sides in the base (which is only a triangle for 4 anchors) and with those two points plus the point on top of the pyramid (the last anchor) you get each of the pyramid triangles.

Then for the base, any 3 points within the base will serve to define the plane.

> Anyhow, this function needs a big battery of tests since it's hard to follow and reason about.

100% agreed. Can you help me with that my pointing where the unit tests are and how they are run?
I can't find them, but this PR should definitely come with some tests, I think.
Once we get all the tests for the pyramid printing space we can try to generalize further and remove the assumption of the pyramid shape and the last anchor being on top. 
Probably M669 could have some validations and return some errors if certain configurations are not supported. 

My plan was to next move from pyramid to prism + pyramid (pyramid being a special case where the prism on the bottom has height=0) and then also allow prism only too, without the need for a top of the pyramid anchor.
We could generalize further, to any volume without anchors in the middle, I think (even the most generalized case would need validations in M669, I think), but what I'm looking for is just a prism.
A quadrangular prism, well, rather pseudo-quadrangular, but octogonal, because I want 2 anchors on each corner to open the door for the extra 3 degrees of freedom.

Well, I shouldn't get too ahead, one step at a time. 
Just optionally moving from a triangular pyramid to a quadrangular pyramid would be a really nice first step in my opinion, since most rooms are squared (or kind of) in my experience.
Like, one anchor in each corner of the room/closet kind of thing.
But, yeah, eventually I just want a squared prism as printing space, just like more traditional printers.

Sorry for the long rant.

Reply:
> if the two additional points are on the same side of that bisection of space, it returns true and otherwise it returns false.
Correct.

> Can you help me with that my pointing where the unit tests are and how they are run?
Sadly I don't think this repo has established much in terms of unit testing. Whenever I've implemented tricky functions, I've made separate repos for developing those functions in isolation. Eg https://gitlab.com/tobben/hangprinter-forward-transform and https://gitlab.com/tobben/hangprinter-flex-compensation.

The linc-repo has better testing structure, and proper data types for points/vertices. I suggest extracting much of what you need from that repo. Even some of the geometrical functions you need are already there. https://gitlab.com/hangprinter/linc

If you make a separate development repo and point me to it I could help you extract some of the relevant stuff from linc.

> Once we get all the tests for the pyramid printing space we can try to generalize further and remove the assumption of the pyramid shape and the last anchor being on top.
Sounds like a solid plan.